spirv_cross.cpp

```diff
@@ -2527,6 +2530,14 @@ vector<string> Compiler::get_entry_points() const
 	return entries;
 }
 
+unordered_map<string, string> Compiler::get_entry_point_name_map() const
```

Comment:
Can this be called before compile()?

Reply:
It might be easier to just query the remapped name for a particular entry point (as the common case is single entry point), and have an interface like string `Compiler::get_entry_point_remapped_name(const string &entry)`.

Reply:
No. The remapping is part of compile().

I implemented the mapping per the request in issue #316. I don't mind either way.

Reply:
Ok, it's probably okay, but this behavior needs to be documented I think. We can enforce this in runtime by throwing an exception if this is called before compile().

I still prefer returning a single string here though. Returning an unordered_map for this seems a bit too much.

Reply:
I don't really have a preference for #316, either `get_entry_point_name_map` or `get_entry_point_remapped_name` are fine IMO

Reply:
Once renaming happens...the existing `get_entry_points()` function will return a list of the **_new_** names (i.e.. in the most common case it will contain `main0`). So the app has no way of getting a list of the **_old_** names...which are the ones the app is likely expecting to be familiar with already.

So...one potential issue with a single-string query-answer function is the app knowing what to use as the look-up key?

1. Would we use the original name as the key...effectively expecting the app to somehow know what the SPIR-V entry points were before renaming...without being able to ask the compiler for that info?

2. Or would the app use `get_entry_points()` to get the new names...and then  cycle through them to find which one mapped to an old name? That seems a bit backwards to me.

Reply:
Regarding the question of timing...

We wouldn't want the renaming to occur during parsing...because that is backend-dependent...and would break the design goal of being able to reuse parsed results to output multiple backend languages.

After parsing but before compiling, `get_entry_point_name_map()` will return a mapping of:

    original names -> original names

After compiling, it will return a mapping of:

    original names -> new names

I have documented this now.

Reply:
I see the API flow as something like this:

- get_entry_points() is generally something you do before compile(), because after all, you need to figure out which entry points you want to use.

- Then, you call compile(), and the entry point names are potentially changed in MSL.

- Before you pass this MSL to the compiler, you can remap the entry point you compiled using get_entry_point_name_mapping(), to see what your actual entry point is.


It should be possible for get_entry_points() to always return the original names I think, it could just return "orig_name". The entry points are a property of the SPIR-V module itself, and I don't think it needs to change like this. The API should always accept original names for entry points, except that the "remapped name" is something which happens in the MSL output, but only there.

Reply:
It might be better to throw an exception if get_entry_point_name_mapping is called before compile() imo. API users will call it, get some sensible result, and be surprised when they get the wrong result. Wrong API use should have an unignorable error.

Reply:
I have changed `get_entry_points()` and `get_entry_point(name)` functions to use the `orig_name` of the entry point instead of `name`.

I have removed the `get_entry_point_name_mapping()` function and introduced a `get_cleansed_entry_point_name(name)` function.

I could not determine an consistent way of detecting whether the `compile()` function had been run across all backends...so I left the idea of throwing an exception unimplemented...and simply documented the `get_cleansed_entry_point_name(name)` function to indicate what the API user should expect.